README.md
[![support](https://img.shields.io/badge/support-GitHub-white)](https://github.com/sponsors/dr-dimitru)
[![support](https://img.shields.io/badge/support-PayPal-white)](https://paypal.me/veliovgroup)
<a href="https://ostr.io/info/built-by-developers-for-developers">
  <img src="https://ostr.io/apple-touch-icon-60x60.png" height="20">
</a>

# Measure FPS

While working with CSS, HTML, and JavaScript, it's always important to take care of browser rendering in browser DevTools. To find issues with rendering, we are using "Timeline" with "JS Profile" and "Memory" consumption tracking. Those tools are great when you know about the existing problem.

During our development process, we are using FPS-meter, which gives a signal when "Long" frames occur. To see meter action, visit [this link](https://cssbuilder.veliovgroup.com) (*in the left bottom corner*). Play with sliders to see how different CSS effects may slow down the rendering process, `blur` is the most "expensive" CSS3 filter.

This library works on mobile devices. It utilizes `performance.now` to measure time frame and `requestAnimationFrame` to measure rendered frames, both APIs are very efficient and have a minor impact (*for correct FPS measurement*).

## Demo:

- Demo can be found on [this website](https://cssbuilder.veliovgroup.com) (*in the left bottom corner*).

## Drop-in version

Installation is not required, copy-paste script into browser' console:

- Drop-in: [minified version](https://github.com/veliovgroup/fps-meter/blob/master/fps-meter-drop-in.min.js);
- Drop-in: [developer version](https://github.com/veliovgroup/fps-meter/blob/master/fps-meter-drop-in.js);
- Link [to minified file](https://raw.githubusercontent.com/veliovgroup/fps-meter/master/fps-meter-drop-in.min.js);
- HTML Script: `<script type="text/javascript" src="https://raw.githubusercontent.com/veliovgroup/fps-meter/master/fps-meter-drop-in.min.js"></script>`

## Installation

Installation is available via NPM for your favorite build tool and Atmosphere for Meteor.js

### NPM install

```shell
npm i --save fps-m
```

### Meteor add

```shell
meteor add ostrio:fps-meter
```

### NPM require

```js
// require
const FPSMeter = require('fps-m');

// ES6 import:
import FPSMeter from 'fps-m';
```

### Meteor: ES6 Import

```js
import { FPSMeter } from 'meteor/ostrio:fps-meter';
```

### Quick start

```js
(new FPSMeter({ui: true})).start();
```

## API

Render, start, pause, and resume FPS Meter

### `new FPSMeter([opts])`

- `opts` {*Object*}
- `opts.ui` {*Boolean*} - Render small box with current FPS into `body`, default: `false`
- `opts.reactive` {*Boolean*} - Store current FPS as reactive data source, default: `false`

```js
// Quick start:
const fps = new FPSMeter({ui: true, reactive: false});
// fps.start();
// fps.stop();
```

#### Methods

- `FPSMeter#start()` — Use to start measuring FPS. If `{ui: true}` small box with current FPS will be rendered into `body`
- `FPSMeter#stop()` — Use to stop measuring FPS. If `{ui: true}` box with current FPS will be removed from `body`
- `FPSMeter#pause()` — Use to pause measuring FPS. If `{ui: true}` box with current FPS will remain in `body`
- `FPSMeter#resume()` — Use to resume measuring FPS
- `FPSMeter#toggle()` — Use to toggle (pause/resume) measuring FPS

#### Properties

- `FPSMeter#isRunning` {*Boolean*} — `true` if meter is running and `false` when meter is on pause, stopped or has not started yet
- `FPSMeter#template` {*Blaze.View*|*undefined*} — When `{ui: true}` this property holds *Blaze.View* instance, otherwise its `undefined`
- `FPSMeter#element` {*DOMElement*|*undefined*} — When `{ui: true}` this property holds *DOMElement* of FPSMeter UI element `div`, otherwise its `undefined`
- `FPSMeter#fps` {*Number*|*ReactiveVar*} — When `{reactive: false}` it holds a {*Number*} with current FPS. When `{reactive: true}` it is an instance of `{*ReactiveVar*}`. Use `.get()` method to return current FPS. It's reactive data source, and can be used in template:

## Examples

### Pause/Resume by clicking on UI box

```js
const fps = new FPSMeter({ ui: true });
fps.start();
fps.element.addEventListener('click', fps.toggle.bind(fps), { passive: true, capture: false });
```

### Meteor.js usage

```js
const fps = new FPSMeter({ui: false, reactive: true});

Template.currentFPS.helpers({
  currentFPS() {
    return fps.fps.get();
  }
});

Template.currentFPS.events({
  'click [data-start-fps]'(e) {
    e.preventDefault();
    fps.start();
    return false;
  },
  'click [data-stop-fps]'(e) {
    e.preventDefault();
    fps.stop();
    return false;
  }
});
```

```html
<template name="currentFPS">
  <div>{{currentFPS}}</div>
  <button data-start-fps>Start</button>
  <button data-stop-fps>Stop</button>
</template>
```

## Support this project

- [Sponsor via GitHub](https://github.com/sponsors/dr-dimitru)
- [Support via PayPal](https://paypal.me/veliovgroup) — support my open source contributions once or on regular basis
- Use [ostr.io](https://ostr.io) — [Monitoring](https://snmp-monitoring.com), [Analytics](https://ostr.io/info/web-analytics), [WebSec](https://domain-protection.info), [Web-CRON](https://web-cron.info) and [Pre-rendering](https://prerendering.com) for a website


HISTORY.md
For full history see [releases on GitHub](https://github.com/veliovgroup/fps-meter/releases)

fps-npm.js
if (!window.requestAnimationFrame) {
  window.requestAnimationFrame = (() => {
    return window.requestAnimationFrame || window.webkitRequestAnimationFrame || window.mozRequestAnimationFrame || function (callback) {
      window.setTimeout(callback, 1000 / 60);
    };
  })();
}

const getTime = () => {
  return (window.performance && window.performance.now) ? window.performance.now() : +new Date();
};

module.exports = class FPSMeter {
  constructor(opts) {
    this.ui = opts.ui || false;
    this.fps = 0;
    this.isRunning = false;
    this.defaultStyles = 'z-index:999;position:fixed;bottom:0;left:0;padding:10px;font-weight:600;font-style:normal;font-size:12px;font-family:Consolas,Menlo,Monaco,"Lucida Console","Liberation Mono","DejaVu Sans Mono","Bitstream Vera Sans Mono","Courier New",Courier,monospace,sans-serif';
  }

  measure() {
    const time = getTime();
    window.requestAnimationFrame(() => {
      const _fps = Math.round((1 / (getTime() - time)) * 1000);

      this.fps = _fps;

      if (this.ui && this.element) {
        let i = 4 - `${_fps}`.length;
        let pad = '';

        while (i > 0) {
          pad += ' ';
          i--;
        }

        this.text.nodeValue = `${_fps}${pad}fps`;

        switch (false) {
        case !(_fps < 7):
          this.element.style.color = '#FFF';
          this.element.style.backgroundColor = '#FF4500';
          break;
        case !(_fps < 25):
          this.element.style.color = '#FF4500';
          this.element.style.backgroundColor = '#000';
          break;
        case !(_fps < 40):
          this.element.style.color = 'orange';
          this.element.style.backgroundColor = '#000';
          break;
        case !(_fps > 70):
          this.element.style.color = '#0f0';
          this.element.style.backgroundColor = '#000';
          break;
        default:
          this.element.style.color = '#018801';
          this.element.style.backgroundColor = '#000';
          break;
        }
      }

      if (this.isRunning) {
        this.measure();
      }
    });
  }

  start() {
    if (!this.isRunning) {
      this.isRunning = true;

      if (this.ui === true) {
        this.text = document.createTextNode('');
        this.element = document.createElement('div');
        this.element.style = this.defaultStyles;
        this.element.appendChild(this.text);
        document.body.appendChild(this.element);
      }

      this.measure();
    }
  }

  pause() {
    this.isRunning = false;
  }

  resume() {
    if (!this.isRunning) {
      this.isRunning = true;
      this.measure();
    }
  }

  toggle() {
    this.isRunning ? this.pause() : this.resume();
  }

  stop() {
    if (this.isRunning) {
      this.isRunning = false;
      if (this.ui === true && this.element) {
        this.element.remove();
      }
    }
  }
};


fps-meter.js
import { Blaze } from 'meteor/blaze';
import { Template } from 'meteor/templating';
import { ReactiveVar } from 'meteor/reactive-var';

if (!window.requestAnimationFrame) {
  window.requestAnimationFrame = (() => {
    return window.requestAnimationFrame || window.webkitRequestAnimationFrame || window.mozRequestAnimationFrame || function (callback) {
      window.setTimeout(callback, 1000 / 60);
    };
  })();
}

const getTime = () => {
  return (window.performance && window.performance.now) ? window.performance.now() : +new Date();
};

class FPSMeter {
  constructor(opts) {
    this.ui = opts.ui || false;
    this.reactive = opts.reactive || false;

    if (this.reactive === true) {
      this.fps = new ReactiveVar(0);
    } else {
      this.fps = 0;
    }

    this.isRunning = false;
    this.template = null;
  }

  measure() {
    const time = getTime();
    window.requestAnimationFrame(() => {
      const _fps = Math.round((1 / (getTime() - time)) * 1000);

      if (this.reactive === true) {
        this.fps.set(_fps);
      } else {
        this.fps = _fps;
      }

      if (this.ui && this.element) {
        let i = 4 - `${_fps}`.length;
        let pad = '';

        while (i > 0) {
          pad += ' ';
          i--;
        }

        this.text.nodeValue = `${_fps}${pad}fps`;

        switch (false) {
        case !(_fps < 7):
          this.element.className = 'dead';
          break;
        case !(_fps < 25):
          this.element.className = 'danger';
          break;
        case !(_fps < 40):
          this.element.className = 'warn';
          break;
        case !(_fps > 70):
          this.element.className = 'high';
          break;
        default:
          this.element.className = '';
          break;
        }
      }

      if (this.isRunning) {
        this.measure();
      }
    });
  }

  start() {
    if (!this.isRunning) {
      this.isRunning = true;

      if (this.ui === true && Blaze) {
        this.template = Blaze.render(Template.FPSMeter, document.body);
        this.text = document.createTextNode('');
        this.element = this.template.templateInstance().find('#__FPSMeter');
        this.element.appendChild(this.text);
      }

      this.measure();
    }
  }

  pause() {
    this.isRunning = false;
  }

  resume() {
    if (!this.isRunning) {
      this.isRunning = true;
      this.measure();
    }
  }

  toggle() {
    this.isRunning ? this.pause() : this.resume();
  }

  stop() {
    if (this.isRunning) {
      this.isRunning = false;
      if (this.ui === true && Blaze && this.template) {
        Blaze.remove(this.template);
      }
    }
  }
}

export { FPSMeter };


fps-meter-drop-in.js
// Start with:
// (new FPSMeter({ui: true})).start();

if (!window.requestAnimationFrame) {
  window.requestAnimationFrame = (() => {
    return window.requestAnimationFrame || window.webkitRequestAnimationFrame || window.mozRequestAnimationFrame || function (callback) {
      window.setTimeout(callback, 1000 / 60);
    };
  })();
}

const getTime = () => {
  return (window.performance && window.performance.now) ? window.performance.now() : +new Date();
};

class FPSMeter {
  constructor(opts) {
    this.ui = opts.ui || false;
    this.fps = 0;
    this.isRunning = false;
    this.defaultStyles = 'z-index:999;position:fixed;bottom:0;left:0;padding:10px;font-weight:600;font-style:normal;font-size:12px;font-family:Consolas,Menlo,Monaco,"Lucida Console","Liberation Mono","DejaVu Sans Mono","Bitstream Vera Sans Mono","Courier New",Courier,monospace,sans-serif';
  }

  measure() {
    const time = getTime();
    window.requestAnimationFrame(() => {
      const _fps = Math.round((1 / (getTime() - time)) * 1000);

      this.fps = _fps;

      if (this.ui && this.element) {
        let i = 4 - `${_fps}`.length;
        let pad = '';

        while (i > 0) {
          pad += ' ';
          i--;
        }

        this.text.nodeValue = `${_fps}${pad}fps`;

        switch (false) {
        case !(_fps < 7):
          this.element.style.color = '#FFF';
          this.element.style.backgroundColor = '#FF4500';
          break;
        case !(_fps < 25):
          this.element.style.color = '#FF4500';
          this.element.style.backgroundColor = '#000';
          break;
        case !(_fps < 40):
          this.element.style.color = 'orange';
          this.element.style.backgroundColor = '#000';
          break;
        case !(_fps > 70):
          this.element.style.color = '#0f0';
          this.element.style.backgroundColor = '#000';
          break;
        default:
          this.element.style.color = '#018801';
          this.element.style.backgroundColor = '#000';
          break;
        }
      }

      if (this.isRunning) {
        this.measure();
      }
    });
  }

  start() {
    if (!this.isRunning) {
      this.isRunning = true;

      if (this.ui === true) {
        this.text = document.createTextNode('');
        this.element = document.createElement('div');
        this.element.style = this.defaultStyles;
        this.element.appendChild(this.text);
        document.body.appendChild(this.element);
      }

      this.measure();
    }
  }

  pause() {
    this.isRunning = false;
  }

  resume() {
    if (!this.isRunning) {
      this.isRunning = true;
      this.measure();
    }
  }

  toggle() {
    this.isRunning ? this.pause() : this.resume();
  }

  stop() {
    if (this.isRunning) {
      this.isRunning = false;
      if (this.ui === true && this.element) {
        this.element.remove();
      }
    }
  }
}
